File: macos/MDreader/util/Titles.swift
import Foundation

enum Titles {
    private static let markdownExtensions: Set<String> = ["md", "markdown", "mdown", "mkd", "mkdown"]

    static func fromPath(_ path: String) -> String {
        if path.isEmpty { return "" }
        let slashIdx = [path.lastIndex(of: "/"), path.lastIndex(of: "\\")].compactMap { $0 }.max()
        let nameStart: String.Index
        if let s = slashIdx {
            nameStart = path.index(after: s)
        } else {
            nameStart = path.startIndex
        }
        let name = String(path[nameStart...])
        guard let dot = name.lastIndex(of: ".") else { return name }
        let dotOffset = name.distance(from: name.startIndex, to: dot)
        if dotOffset <= 0 { return name }
        let ext = String(name[name.index(after: dot)...]).lowercased()
        if markdownExtensions.contains(ext) {
            return String(name[..<dot])
        }
        return name
    }
}

File: android/app/src/main/java/com/mdreader/util/Titles.kt
package com.mdreader.util

/** Pure helpers for deriving human-friendly titles from file paths. */
object Titles {

    private val markdownExtensions = setOf("md", "markdown", "mdown", "mkd", "mkdown")

    /**
     * Returns a display title from a file path or URI path: takes the segment
     * after the last separator and strips a known markdown extension.
     * Non-markdown extensions are preserved.
     */
    fun fromPath(path: String): String {
        if (path.isEmpty()) return ""
        val slash = maxOf(path.lastIndexOf('/'), path.lastIndexOf('\\'))
        val name = if (slash >= 0) path.substring(slash + 1) else path
        val dot = name.lastIndexOf('.')
        if (dot <= 0) return name
        val ext = name.substring(dot + 1).lowercase()
        return if (ext in markdownExtensions) name.substring(0, dot) else name
    }
}
